// react-in-viewport/dist/stories/common/IframeFunctional.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactAspectRatio = _interopRequireDefault(require("react-aspect-ratio"));

var _index = require("../../index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var IframeFunctional = function IframeFunctional(props) {
  var inViewport = props.inViewport,
      src = props.src,
      ratio = props.ratio,
      forwardedRef = props.forwardedRef;

  var _useState = (0, _react.useState)(false),
      loaded = _useState[0],
      setLoaded = _useState[1];

  (0, _react.useEffect)(function () {
    if (inViewport && !loaded) {
      setLoaded(true);
    }
  }, [inViewport, loaded]);
  var Component = loaded ? 'iframe' : 'div';
  var componentProps = loaded ? {
    src: src,
    frameBorder: 0
  } : {};
  return _react["default"].createElement(_reactAspectRatio["default"], {
    ratio: ratio,
    style: {
      marginBottom: '200px',
      backgroundColor: 'rgba(0,0,0,.12)'
    },
    ref: forwardedRef
  }, _react["default"].createElement(Component, componentProps));
};

var LazyIframe = (0, _index.handleViewport)((0, _react.memo)(IframeFunctional), {}, {
  disconnectOnLeave: true
});
var _default = LazyIframe;
exports["default"] = _default;
