// react-in-viewport/dist/stories/common/Image.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactAspectRatio = _interopRequireDefault(require("react-aspect-ratio"));

var _constants = require("./constants");

var _index = require("../../index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var DUMMY_IMAGE_SRC = 'https://www.gstatic.com/psa/static/1.gif';

var ImageObject =
/*#__PURE__*/
function (_PureComponent) {
  _inheritsLoose(ImageObject, _PureComponent);

  // eslint-disable-line
  function ImageObject(props) {
    var _this;

    _this = _PureComponent.call(this, props) || this;

    _defineProperty(_assertThisInitialized(_this), "loadImage", function (src) {
      var img = new Image(); // eslint-disable-line

      _this.setState({
        status: _constants.LOADING
      });

      img.onload = function () {
        _this.setState({
          src: src,
          status: _constants.LOADED
        });
      };

      img.alt = 'demo';
      img.src = src;
    });

    _defineProperty(_assertThisInitialized(_this), "getBackgroundColor", function () {
      var status = _this.state.status;

      switch (status) {
        case _constants.LOADING:
          return 'rgba(0,0,0,.32)';

        case _constants.LOADED:
          return 'rgba(0,0,0,.50)';

        case _constants.INIT:
        default:
          return 'rgba(0,0,0,.12)';
      }
    });

    _this.state = {
      src: DUMMY_IMAGE_SRC,
      status: _constants.INIT
    };
    return _this;
  }

  var _proto = ImageObject.prototype;

  _proto.componentDidMount = function componentDidMount() {
    if (this.props.inViewport) {
      this.loadImage(this.props.src);
    }
  };

  _proto.componentDidUpdate = function componentDidUpdate() {
    if (this.props.inViewport && this.state.status === _constants.INIT) {
      this.loadImage(this.props.src);
    }
  };

  _proto.render = function render() {
    var status = this.state.status;
    return _react["default"].createElement(_reactAspectRatio["default"], {
      ratio: this.props.ratio,
      style: {
        transitionDuration: '300ms',
        maxWidth: '400px',
        marginBottom: '200px',
        backgroundColor: this.getBackgroundColor()
      }
    }, _react["default"].createElement("img", {
      src: this.state.src,
      alt: "demo"
    }));
  };

  return ImageObject;
}(_react.PureComponent);

var LazyImage = (0, _index.handleViewport)(ImageObject, {}, {
  disconnectOnLeave: true
});
var _default = LazyImage;
exports["default"] = _default;
