// react-in-viewport/dist/stories/common/ImageFunctional.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactAspectRatio = _interopRequireDefault(require("react-aspect-ratio"));

var _index = require("../../index");

var _constants = require("./constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var DUMMY_IMAGE_SRC = 'https://www.gstatic.com/psa/static/1.gif';

var ImageObject = function ImageObject(props) {
  var originalSrc = props.src,
      ratio = props.ratio,
      forwardedRef = props.forwardedRef,
      inViewport = props.inViewport;

  var _useState = (0, _react.useState)(DUMMY_IMAGE_SRC),
      src = _useState[0],
      setSrc = _useState[1];

  var _useState2 = (0, _react.useState)(_constants.INIT),
      status = _useState2[0],
      setStatus = _useState2[1];

  var loadImage = function loadImage(imageSrc) {
    var img = new Image(); // eslint-disable-line

    setStatus(_constants.LOADING);

    img.onload = function () {
      setSrc(imageSrc);
      setStatus(_constants.LOADED);
    };

    img.alt = 'demo';
    img.src = imageSrc;
  };

  (0, _react.useEffect)(function () {
    if (inViewport && status === _constants.INIT) {
      loadImage(originalSrc);
    }
  }, [inViewport, status]);

  var getBackgroundColor = function getBackgroundColor() {
    switch (status) {
      case _constants.LOADING:
        return 'rgba(0,0,0,.32)';

      case _constants.LOADED:
        return 'rgba(0,0,0,.50)';

      case _constants.INIT:
      default:
        return 'rgba(0,0,0,.12)';
    }
  };

  return _react["default"].createElement(_reactAspectRatio["default"], {
    ratio: ratio,
    style: {
      transitionDuration: '300ms',
      maxWidth: '400px',
      marginBottom: '200px',
      backgroundColor: getBackgroundColor()
    },
    ref: forwardedRef
  }, _react["default"].createElement("img", {
    src: src,
    alt: "demo"
  }));
};

var LazyImage = (0, _index.handleViewport)(ImageObject, {}, {
  disconnectOnLeave: true
});
var _default = LazyImage;
exports["default"] = _default;
